### ISSUE TITLE
[rustdoc] `error : too many file operands` when experimental feature `-w` is passed without parameters

### ISSUE THREAD

**Issue by NovaliX-Dev** [rust-lang/rust#132718]
When i try to run this command : `cargo +nightly rustdoc -- -Z unstable-options -w`

Rustdoc returns this error : 
```
error: too many file operands
```

Which confuses me. I mean, i didn't passed any file format, so how can there be too many of them ?

We should have a clearer error, something like this : 
```
error: no output format given after `-w`
```

**Comment by fmease** [comment]
While `cargo rustdoc` comes from a different repository (namely https://github.com/rust-lang/cargo), the `rustdoc` binary *does* bear the blame here:

So, assuming the package contains a library crate, Cargo executes the following (manually adjusted for legibility):

```sh
rustdoc --edition=2021 --crate-type lib --crate-name "$CRATE_NAME" src/lib.rs \
    -o "$CARGO_TARGET_DIR/doc" --check-cfg 'cfg(docsrs)' --check-cfg 'cfg(feature, values())' \
    --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --diagnostic-width=192 \
    -C metadata=$DATA -L "dependency=$CARGO_TARGET_DIR/debug/deps" \
    -Z unstable-options -w --crate-version "$CRATE_VERSION"
```

Now, as you can see, Cargo passes extra arguments after `-w` which should've been fine but here's the kicker: For some reason, `rustdoc`/`getopts` tries to pass the argument for `-w` *greedily* or *verbatim*. Meaning that it considers the string `--crate-version` to be the argument for `-w` (!) instead of a separate flag (and `-w`'s arg to be none or empty). As a result of a that, the crate version `"$CRATE_VERSION"` (e.g., `0.1.0`) gets interpreted as a stray file operand (top-level argument). Hence the error.

If possible (`getopts` is quite primitive), the fix would be to not so greedily parse the argument of `-w`. This should lead to the error changing to `error: Argument to option 'w' missing`.

**Comment by fmease** [comment]
> For some reason, rustdoc/getopts tries to parse the argument for -w greedily or verbatim

I've just noticed this seems to affect *all* "argument-taking" flags, e.g. `--crate-version --verbose` literally sets the crate version to `--verbose`. Oh no, `getopts` >:(

**Comment by fmease** [comment]
Switching `getopt`'s [parsing style](https://docs.rs/getopts/latest/getopts/struct.Options.html#method.parsing_style) from [`FloatingFrees`](https://docs.rs/getopts/latest/getopts/enum.ParsingStyle.html#variant.FloatingFrees) to [`StopAtFirstFree`](https://docs.rs/getopts/latest/getopts/enum.ParsingStyle.html#variant.StopAtFirstFree) is not what we want.

Ah, I think we want to update all references of  `.optopt(…)` to `.optflagopt(…)`. One needs to double-check what other consequences that would have and if it's would be a breaking change (accept fewer or more program invocations).

**Comment by GuillaumeGomez** [comment]
Seems like a fix is required in `getopts`. Something like:

```rust
enum ArgumentStatus<T> {
    // `T` is the value, so if `None`, then we can emit the right error message
    Present(T),
    NotPresent.
}
```

maybe?

**Comment by NovaliX-Dev** [comment]
For single-use arguments that seems to be something that can be tested.

We still need to take into account multiple-use arguments (arguments callable multiple times) though. I don't know if getopts support them.

**Comment by fmease** [comment]
Fair point regarding "weird paths" (esp. w/ shell variable substitutions).

At the risk of getting slightly off-topic here (since changing rust{,do}c's arg parsing is out of the question), clap for example still supports paths starting with dash(es) if you use `=` (so, `-w --crate-version` would mean opt `-w` \mapsto `Some(None)`, opt `--crate-version` \mapsto `Some(None)` and `-w=--crate-version` would mean opt `-w` \mapsto `Some("--crate-versions")`) but I guess there are many conventions.

Regarding the ordering we should probably open an issue or better yet a PR over at the cargo repo. That'll improve the diagnostic in some more cases at least.

**Comment by lolbinarycat** [comment]
clap's behavior sounds like a cve waiting to happen due to some script not using `=` when it should.

**Comment by NovaliX-Dev** [comment]
> honestly i don't think this is a bug, this is generally just how unix-y options work.
> 
> if you run `tar -xf --version`, it will attempt to extract an archive named `--version`.
> 
> this is actually a good thing, otherwise you would have all sorts of problems with programs not handling weirdly named files correctly (well, even more than we already do).
> 
> i think the weird thing here is that `cargo rustdoc` puts the user-provided flags in the middle of the automatically generated flags, instead of after them.

That's probably going to be off topic, but I found the behavior you're describing weird. I mean, from my perspective, when you start a part of a command (by part I mean all the sequences of chars which don't contains a whitespace) by a dash "-", 99% of the time you consider that part as an argument and not a value to another argument. 
As for the weird named files, I think quoting them is a better idea rather than just accepting the whole part without checking if it's an argument or not.

I'll be honest, maybe you know things better than I do, and maybe there are cases I didn't consider. I'm using Linux since only a few months, so probably i'm still very inexperienced. If you have any articles / arguments / links explaining why the approach you're describing is better than mine, I would like to hear them.

But still, my way seems so much more logical to me.

I hope my comment doesn't sound angry or anything alike, I'm conscious that sometimes I can write my thoughts in a "raw" way.

**Comment by lolbinarycat** [comment]
> I think quoting them is a better idea rather than just accepting the whole part without checking if it's an argument or not.

that doesn't work, quoting is a shell construct, once the arguments get to the program, it can't tell if it was quoted or not.

at least on unix that is, on windows the program is responsible for the entirety of argument parsing, so what you describe would work there.

**Comment by lolbinarycat** [comment]
also remember, if a flag takes an parameter separated by whitespace, that parameter is *mandatory*, so the situation you describe (and indeed, the behavior of clap), will only ever cause extra errors.
